### PR TITLE
Implement createObservable for PreparedGetObject

### PR DESCRIPTION
**Pull request by geralt-encore**
I am not sure about my integration tests

### CODE REVIEW

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java [hunk 1]
```diff
@@ -199,4 +202,55 @@ public void nestedTransaction() {
...
+                        .where(UserTableMeta.COLUMN_EMAIL + "=?")
+                        .whereArgs(expectedUser.email())
+                        .build())
+                .prepare()
+                .createObservable();
+
+        TestSubscriber<User> testSubscriber = new TestSubscriber<User>();
+        userObservable.subscribe(testSubscriber);
+
+        testSubscriber.awaitTerminalEvent(500, TimeUnit.MILLISECONDS);
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java]: static import for milliseconds please :)

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/operations/get/PreparedGetObjectTest.java [hunk 2]
```diff
@@ -122,6 +192,41 @@ public void shouldThrowExceptionIfNoTypeMappingWasFoundWithoutAccessingDbWithQue
...
+                    .withQuery(Query.builder().table("test_table").build())
+                    .prepare()
+                    .createObservable()
+                    .subscribe(testSubscriber);
+
+            testSubscriber.awaitTerminalEvent();
+            testSubscriber.assertNoValues();
+            assertThat(testSubscriber.getOnErrorEvents().get(0))
+                    .isInstanceOf(StorIOException.class)
+                    .hasCauseInstanceOf(IllegalStateException.class);
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/operations/get/PreparedGetObjectTest.java]: Let's also check exception message here

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/operations/get/PreparedGetObjectTest.java [hunk 3]
```diff
@@ -153,6 +258,37 @@ public void shouldThrowExceptionIfNoTypeMappingWasFoundWithoutAccessingDbWithRaw
...
+                    .withQuery(RawQuery.builder().query("test query").build())
+                    .prepare()
+                    .createObservable()
+                    .subscribe(testSubscriber);
+
+            testSubscriber.awaitTerminalEvent();
+            testSubscriber.assertNoValues();
+            assertThat(testSubscriber.getOnErrorEvents().get(0))
+                    .isInstanceOf(StorIOException.class)
+                    .hasCauseInstanceOf(IllegalStateException.class);
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/operations/get/PreparedGetObjectTest.java]: Exception message

**Comment by artem-zinnatullin** [comment]
Looks good! Just few comments.

**Comment by artem-zinnatullin** [comment]
@nikitin-da PTAL :)

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/operations/get/PreparedGetObjectTest.java [hunk 4]
```diff
@@ -42,6 +50,22 @@ public void shouldGetByQueryWithoutTypeMappingBlocking() {
         }
 
         @Test
+        public void shouldGetByQueryWithoutTypeMappingAsObservable() {
```
Comment (nikitin-da) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/operations/get/PreparedGetObjectTest.java]: Maybe `shouldGetObjectByQueryWithoutTypeMappingAsObservable` will be better=)

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/operations/get/PreparedGetObjectTest.java [hunk 5]
```diff
@@ -195,6 +331,26 @@ public void executeAsBlockingShouldThrowExceptionIfNoQueryWasSet() {
...
+                    (Query) null,
+                    (GetResolver<Object>) mock(GetResolver.class)
+            );
+
+            try {
+                //noinspection ResourceType
+                preparedGetOfObject.createObservable();
+                failBecauseExceptionWasNotThrown(IllegalStateException.class);
+            } catch (IllegalStateException expected) {
+                assertThat(expected).hasMessage("Please specify query");
```
Comment (nikitin-da) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/operations/get/PreparedGetObjectTest.java]: Mainstream comment about exceptions checking. There is no cause verification.

**Comment by nikitin-da** [comment]
:+1: LGTM

**Comment by artem-zinnatullin** [comment]
Would be nice to also have integration tests to see that it reacts on changes of tables from query and emits new result!

**Comment by geralt-encore** [comment]
Ok, I will get my hands on it today in the evening

**Comment by artem-zinnatullin** [comment]
Thanks!

**Comment by geralt-encore** [comment]
@artem-zinnatullin check this out!

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/design/GetOperationDesignTest.java [hunk 6]
```diff
@@ -153,4 +153,33 @@ public void getObjectBlockingWithRawQueryBlocking() {
...
+                        .where("email = ?")
+                        .whereArgs("artem.zinnatullin@gmail.com")
+                        .build())
+                .withGetResolver(UserTableMeta.GET_RESOLVER)
+                .prepare()
+                .createObservable();
+    }
+
+    @Test
+    public void getObjectBlockingWithRawQueryObservable() {
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/design/GetOperationDesignTest.java]: Unneeded `Blocking` in test name

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java [hunk 7]
```diff
@@ -199,4 +202,116 @@ public void nestedTransaction() {
...
+
+        testSubscriber.awaitTerminalEvent(500, MILLISECONDS);
+        testSubscriber.assertValueCount(1);
+        testSubscriber.assertValue(null);
+        testSubscriber.assertNoErrors();
+        testSubscriber.unsubscribe();
+    }
+
+    @Test
+    public void queryOneExistedObjectTableUpdate() {
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java]: Nice test!

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java [hunk 8]
```diff
@@ -199,4 +202,116 @@ public void nestedTransaction() {
...
+        testSubscriber.assertValueCount(1);
+        testSubscriber.assertValue(null);
+        testSubscriber.assertNoErrors();
+
+        putUserBlocking(expectedUser);
+
+        testSubscriber.awaitTerminalEvent(500, MILLISECONDS);
+        testSubscriber.assertValueCount(2);
+        testSubscriber.assertValues(null, expectedUser);
+        testSubscriber.assertNoErrors();
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java]: `testSubscriber.unsubscribe()`

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java [hunk 9]
```diff
@@ -199,4 +202,116 @@ public void nestedTransaction() {
...
+        testSubscriber.assertValueCount(1);
+        testSubscriber.assertValue(null);
+        testSubscriber.assertNoErrors();
+
+        putUserBlocking();
+
+        testSubscriber.awaitTerminalEvent(500, MILLISECONDS);
+        testSubscriber.assertValueCount(2);
+        testSubscriber.assertValues(null, null);
+        testSubscriber.assertNoErrors();
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java]: `testSubscriber.unsubscribe()`

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java [hunk 10]
```diff
@@ -199,4 +202,116 @@ public void nestedTransaction() {
...
+                        .where(UserTableMeta.COLUMN_EMAIL + "=?")
+                        .whereArgs("some arg")
+                        .build())
+                .prepare()
+                .createObservable();
+
+        TestSubscriber<User> testSubscriber = new TestSubscriber<User>();
+        userObservable.subscribe(testSubscriber);
+
+        testSubscriber.awaitTerminalEvent(500, MILLISECONDS);
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java]: Not needed here

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java [hunk 11]
```diff
@@ -199,4 +202,116 @@ public void nestedTransaction() {
...
+        testSubscriber.assertValueCount(1);
+        testSubscriber.assertValue(null);
+        testSubscriber.assertNoErrors();
+
+        putUserBlocking();
+
+        testSubscriber.awaitTerminalEvent(500, MILLISECONDS);
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java]: And here

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java [hunk 12]
```diff
@@ -199,4 +202,116 @@ public void nestedTransaction() {
...
+                        .where(UserTableMeta.COLUMN_EMAIL + "=?")
+                        .whereArgs(expectedUser.email())
+                        .build())
+                .prepare()
+                .createObservable();
+
+        TestSubscriber<User> testSubscriber = new TestSubscriber<User>();
+        userObservable.subscribe(testSubscriber);
+
+        testSubscriber.awaitTerminalEvent(500, MILLISECONDS);
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java]: Not needed

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java [hunk 13]
```diff
@@ -199,4 +202,116 @@ public void nestedTransaction() {
...
+        testSubscriber.assertValueCount(1);
+        testSubscriber.assertValue(null);
+        testSubscriber.assertNoErrors();
+
+        putUserBlocking(expectedUser);
+
+        testSubscriber.awaitTerminalEvent(500, MILLISECONDS);
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java]: Not needed

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java [hunk 14]
```diff
@@ -199,4 +202,116 @@ public void nestedTransaction() {
...
+                        .where(UserTableMeta.COLUMN_EMAIL + "=?")
+                        .whereArgs("some arg")
+                        .build())
+                .prepare()
+                .createObservable();
+
+        TestSubscriber<User> testSubscriber = new TestSubscriber<User>();
+        userObservable.subscribe(testSubscriber);
+
+        testSubscriber.awaitTerminalEvent(500, MILLISECONDS);
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java]: Not needed

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java [hunk 15]
```diff
@@ -199,4 +202,116 @@ public void nestedTransaction() {
...
+                        .where(UserTableMeta.COLUMN_EMAIL + "=?")
+                        .whereArgs(expectedUser.email())
+                        .build())
+                .prepare()
+                .createObservable();
+
+        TestSubscriber<User> testSubscriber = new TestSubscriber<User>();
+        userObservable.subscribe(testSubscriber);
+
+        testSubscriber.awaitTerminalEvent(500, MILLISECONDS);
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java]: Not needed

**Comment by artem-zinnatullin** [comment]
Checked :dancers:

**Comment by geralt-encore** [comment]
But without this ` testSubscriber.awaitTerminalEvent(500, MILLISECONDS)` my tests are failing. If I could somehow change `Schedulers.io()` to `Schedulers.immediate()` i would fix problem, but I couldn't find easy solution for it

**Comment by artem-zinnatullin** [comment]
`testSubscriber.awaitTerminalEvent` is just an attempt to wait, tests can fail on slow computer.

Better to do `take(1)`, etc to explicitly set expected number of items and then do `awaitTerminalEvent()` with huge timeout just to break the test if something gone wrong!

**Comment by geralt-encore** [comment]
Will it be something like that?

      @Test
    public void queryOneExistedObjectObservable() {
        final List<User> users = putUsersBlocking(3);
        final User expectedUser = users.get(0);

        final Observable<User> userObservable = storIOSQLite
                .get()
                .object(User.class)
                .withQuery(Query.builder()
                        .table(UserTableMeta.TABLE)
                        .where(UserTableMeta.COLUMN_EMAIL + "=?")
                        .whereArgs(expectedUser.email())
                        .build())
                .prepare()
                .createObservable()
                .take(1);

        TestSubscriber<User> testSubscriber = new TestSubscriber<User>();
        userObservable.subscribe(testSubscriber);

        testSubscriber.awaitTerminalEvent(5, SECONDS);
        testSubscriber.assertValueCount(1);
        testSubscriber.assertValue(expectedUser);
        testSubscriber.assertNoErrors();
        testSubscriber.unsubscribe();
    }

**Comment by artem-zinnatullin** [comment]
Yep, looks good, you can remove `assertValueCount()` because `assertValue()` checks values count :)

**Comment by geralt-encore** [comment]
Tests are fixed now

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java [hunk 16]
```diff
@@ -199,4 +202,116 @@ public void nestedTransaction() {
...
+                .createObservable()
+                .take(1);
+
+        TestSubscriber<User> testSubscriber = new TestSubscriber<User>();
+        userObservable.subscribe(testSubscriber);
+
+        testSubscriber.awaitTerminalEvent(5, SECONDS);
+        testSubscriber.assertValue(expectedUser);
+        testSubscriber.assertNoErrors();
+        testSubscriber.unsubscribe();
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java]: Not needed but I think I'll merge it anyway :)

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java [hunk 17]
```diff
@@ -199,4 +202,116 @@ public void nestedTransaction() {
...
+                .createObservable()
+                .take(1);
+
+        TestSubscriber<User> testSubscriber = new TestSubscriber<User>();
+        userObservable.subscribe(testSubscriber);
+
+        testSubscriber.awaitTerminalEvent(5, SECONDS);
+        testSubscriber.assertValue(null);
+        testSubscriber.assertNoErrors();
+        testSubscriber.unsubscribe();
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java]: Same as above

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java [hunk 18]
```diff
@@ -199,4 +202,116 @@ public void nestedTransaction() {
...
+        testSubscriber.awaitTerminalEvent(5, SECONDS);
+        testSubscriber.assertValue(null);
+        testSubscriber.assertNoErrors();
+
+        putUserBlocking(expectedUser);
+
+        testSubscriber.awaitTerminalEvent(5, SECONDS);
+        testSubscriber.assertValues(null, expectedUser);
+        testSubscriber.assertNoErrors();
+        testSubscriber.unsubscribe();
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java]: same as ^

File: storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java [hunk 19]
```diff
@@ -199,4 +202,116 @@ public void nestedTransaction() {
...
+        testSubscriber.awaitTerminalEvent(5, SECONDS);
+        testSubscriber.assertValue(null);
+        testSubscriber.assertNoErrors();
+
+        putUserBlocking();
+
+        testSubscriber.awaitTerminalEvent(5, SECONDS);
+        testSubscriber.assertValues(null, null);
+        testSubscriber.assertNoErrors();
+        testSubscriber.unsubscribe();
```
Comment (artem-zinnatullin) [inline storio-sqlite/src/test/java/com/pushtorefresh/storio/sqlite/integration/RxQueryTest.java]: ^

**Comment by artem-zinnatullin** [comment]
:+1: thanks again!

**Comment by geralt-encore** [comment]
I'll try to finish Rx part for content resolver by the end of this week. Could you explain why calling `unsubscribe` for `TestSubscriber` is unnecessary? Sorry for a huge amount of defects in this PR =(

**Comment by artem-zinnatullin** [comment]
@geralt-encore no problems! You help us a lot (I'm busy as always)! Also, you learned some Rx and that's great :)

**Pull request merged**